### PR TITLE
fix: Prevent SQL injection in Data Query by using parameterized queries

### PR DESCRIPTION
**Pull request by WashizuRyo**
### Summary
This PR fixes a SQL injection vulnerability in the Data Query node by implementing parameterized queries. User-controlled inputs are now safely passed as parameters ($1, $2, etc.) instead of being directly interpolated into SQL strings. LLM-generated content remains unparameterized to support use cases where LLM generates complete SQL queries.

### Related Issue
N/A

### Changes
- Implement parameterized queries using PostgreSQL's $1, $2 syntax for user-controlled inputs
- Extract parameterizeQuery as a pure function for better testability
- Separate value resolution from parameterization logic (pure, sync)
- User-controlled inputs (text, trigger, action, appEntry) are parameterized for SQL injection prevention
- LLM-generated content (textGeneration, contentGeneration) remains unparameterized to allow dynamic SQL generation
- Use consistent naming: replaceKeyword for {{...}}, placeholder for $1
- Add comprehensive unit tests for parameterizeQuery function (mock-free)

### Testing
Unit tests cover:
SQL injection prevention with malicious input ('); DROP TABLE users; --)
Duplicate keyword handling (reuses same placeholder index)
Multiple different keywords with sequential placeholders
Empty string values
Unquoted keywords
Queries without keywords

### Other Information
Known limitation: LIKE patterns with placeholders inside string literals (e.g., '%{{...}}%') do not work correctly. This is a known issue and will be addressed separately if needed.

___

### **PR Type**
Bug fix

___

### **Description**
- Implement parameterized SQL queries for user-controlled inputs to prevent SQL injection attacks

- Distinguish between user inputs (parameterized) and LLM-generated content (direct substitution)

- Add `ResolvedQuery` object returning parameterized query, display query, and parameter values

- Create `createParamHelper()` utility to track parameter indices and reuse for duplicate placeholders

- Add comprehensive test suite covering SQL injection, duplicates, and edge cases

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["User Input<br/>Text/AppEntry/Trigger"] -->|"Parameterized<br/>Safe"| B["ResolvedQuery"]
  C["LLM Generated<br/>Content"] -->|"Direct Substitution<br/>Flexible"| B
  B --> D["parameterizedQuery<br/>$1, $2..."]
  B --> E["displayQuery<br/>Actual values"]
  B --> F["values<br/>Parameter array"]
  D -->|"Execute"| G["PostgreSQL Client"]
  E -->|"Display"| H["UI Output"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>execute-data-query.test.ts</strong><dd><code>Add unit tests for parameterized query resolution</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/giselle/src/operations/execute-data-query.test.ts

<ul><li>Added comprehensive test suite for <code>resolveQuery()</code> function with 7 test <br>cases<br> <li> Tests cover SQL injection prevention, duplicate placeholder handling, <br>multiple placeholders, empty strings, unquoted placeholders, and <br>queries without placeholders<br> <li> Uses test fixtures to create text nodes and generations for isolated <br>testing</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2673/files#diff-fcfd69fd91b3a245f583b7d59185875af7e3831f450b93101222216c026ca548">+181/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>execute-data-query.ts</strong><dd><code>Implement parameterized queries with SQL injection protection</code></dd></summary>
<hr>

packages/giselle/src/operations/execute-data-query.ts

<ul><li>Refactored <code>resolveQuery()</code> to return <code>ResolvedQuery</code> object with <br><code>parameterizedQuery</code>, <code>displayQuery</code>, and <code>values</code> properties<br> <li> Added <code>createParamHelper()</code> utility function to track parameter indices <br>and reuse same index for duplicate placeholders<br> <li> Implemented <code>replaceWithParam()</code> function to safely replace <br>user-controlled placeholders with parameterized markers ($1, $2, etc.)<br> <li> Distinguished between user-controlled inputs (text, appEntry, trigger, <br>action) using parameterization and LLM-generated content <br>(textGeneration, contentGeneration) using direct substitution<br> <li> Updated query execution to use parameterized query with values array <br>instead of raw query string<br> <li> Updated output to use <code>displayQuery</code> for UI display instead of <br>parameterized query<br> <li> Handles both quoted (<code>'{{...}}'</code>) and unquoted (<code>{{...}}</code>) placeholders <br>correctly</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2673/files#diff-954e6c1997123a6e0e44bf8df68d96ea202ebc1c4500abc2004622b41278344b">+91/-13</a>&nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

* **Tests**
  * Added unit tests validating query parameterization behavior across repeated, quoted, unquoted, and empty placeholders and the resulting display formatting.

* **Improvements**
  * Query execution now uses positional parameterization for safer database calls and emits a separate, user-facing formatted query for display, improving reliability and preventing injection-prone substitutions.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

---

> [!NOTE]
> Strengthens data query safety and clarity by separating value resolution from parameterization and executing with placeholders.
> 
> - Extracts pure `parameterizeQuery` to convert `{{nodeId:outputId}}` to `$1`, `$2`, etc., returning `parameterizedQuery`, `displayQuery`, and `values`
> - Refactors `resolveQuery` to collect user inputs vs. LLM-generated SQL, applying direct substitutions first, then parameterizing user inputs
> - Updates execution to call `client.query(parameterizedQuery, values)` and to output `displayQuery` for UI
> - Adds concise tests for `parameterizeQuery` covering quoted/unquoted, duplicates, multiple placeholders, empty values, and no-placeholders
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e7be80f998571a725f5ab9c26bdc02080bc72aa4. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: packages/giselle/src/operations/execute-data-query.ts [hunk 1]
```diff
@@ -276,6 +315,33 @@ async function resolveQuery(
...
+
+			// Parameterized query: replace with $1, $2, etc.
+			// Handle quoted placeholder: '{{...}}' → $1 (removes surrounding quotes)
+			parameterizedQuery = parameterizedQuery.replaceAll(
+				quotedReplaceKeyword,
+				placeholder,
+			);
+			// Handle unquoted placeholder: {{...}} → $1
+			parameterizedQuery = parameterizedQuery.replaceAll(
+				replaceKeyword,
```
[the remark this answers is not in the document]
Comment (WashizuRyo) [reply packages/giselle/src/operations/execute-data-query.ts]: Thank you for the detailed analysis.

This is an intentional design decision:

- **Edge case** - Embedding placeholders in string literals (`'%{{id}}%'`) is an unusual pattern
- **Workaround exists** - Users can include `%` in the node value itself: `WHERE name LIKE '{{id}}'` with node value `%john%`
- **Complexity** - Handling this would require regex-based SQL parsing to detect and rewrite patterns like `'%{{...}}%'` or `'prefix-{{...}}-suffix'`, adding significant complexity for a rare use case

If this pattern becomes common in practice, we can revisit and add support for it.

File: packages/giselle/src/operations/execute-data-query.ts [hunk 2]
```diff
@@ -99,7 +99,7 @@ export function executeDataQuery(args: {
 				let result: QueryResult;
 				try {
 					await client.connect();
-					result = await client.query(query);
+					result = await client.query(parameterizedQuery, values);
```
[the remark this answers is not in the document]
Comment (WashizuRyo) [reply packages/giselle/src/operations/execute-data-query.ts:102]: Thank you for the analysis.

I tested this locally and confirmed that `pg` library accepts multi-statement queries even with an empty array parameter, so the parameterized query approach does not cause a regression here.

However, our application currently does not support multi-statement queries at the UI level. We'll consider this separately.

File: packages/giselle/src/operations/execute-data-query.ts [hunk 3]
```diff
@@ -276,6 +315,33 @@ async function resolveQuery(
...
+				placeholder,
+			);
+			// Handle unquoted placeholder: {{...}} → $1
+			parameterizedQuery = parameterizedQuery.replaceAll(
+				replaceKeyword,
+				placeholder,
+			);
+
+			// Display query: replace with actual value (like original behavior)
+			displayQuery = displayQuery.replaceAll(replaceKeyword, stringValue);
```
[the remark this answers is not in the document]
Comment (WashizuRyo) [reply packages/giselle/src/operations/execute-data-query.ts]: The executed query cannot be retrieved from PostgreSQL, and perfectly reconstructing the display query would add significant complexity. We'll leave this as-is for now.

**Comment by WashizuRyo** [comment]
I'll QA testing next week.

**Comment by WashizuRyo** [comment]
Done ✅ https://github.com/giselles-ai/giselle/pull/2673#issuecomment-3788476761
